operators/tool_tracking_postprocessor/tool_tracking_postprocessor.cpp

```diff
@@ -109,118 +90,105 @@ void ToolTrackingPostprocessorOp::compute(InputContext& op_input, OutputContext&
-    const size_t layer_size = output_shape.dimension(0) * output_shape.dimension(1);
-    bool first = true;
-    for (auto& index : visible_classes) {
+    // build a vector with the colors, if more colors are required than specified, repeat the
+    // last color
+    std::vector<float3> colors;
+    for (auto index = 0; index < num_colors_; ++index) {
       const auto& img_color =
-          overlay_img_colors_.get()[std::min(index, uint32_t(overlay_img_colors_.get().size()))];
-      const std::array<float, 3> color{{img_color[0], img_color[1], img_color[2]}};
-      cuda_postprocess(output_shape.dimension(0),
-                       output_shape.dimension(1),
-                       color,
-                       first,
-                       static_cast<float*>(binary_masks_tensor->data()) + index * layer_size,
-                       reinterpret_cast<float4*>(out_data),
-                       cuda_stream_handler_.get_cuda_stream(context.context()));
-      first = false;
+          overlay_img_colors_.get()[std::min(index, int(overlay_img_colors_.get().size()))];
+      colors.push_back(make_float3(img_color[0], img_color[1], img_color[2]));
     }
+
+    CUDA_TRY(cudaMemcpyAsync(dev_colors_,
+                             colors.data(),
+                             num_colors_ * sizeof(float3),
+                             cudaMemcpyHostToDevice,
+                             cuda_stream));
   }
 
-  // pass the CUDA stream to the output message
-  stream_handler_result = cuda_stream_handler_.to_message(out_message_device);
+  // filter coordinates based on probability and create a colored mask from the binary mask
+  cuda_postprocess(probs_tensor->size(),
+                   min_prob_,
+                   reinterpret_cast<const float*>(probs_tensor->data()),
+                   reinterpret_cast<const float2*>(scaled_coords_tensor->data()),
+                   reinterpret_cast<float3*>(out_coords_tensor.value()->pointer()),
+                   mask_shape.dimension(0),
+                   mask_shape.dimension(1),
+                   reinterpret_cast<const float3*>(dev_colors_),
+                   reinterpret_cast<const float*>(binary_masks_tensor->data()),
+                   reinterpret_cast<float4*>(out_mask_tensor.value()->pointer()),
+                   cuda_stream);
 
+  // pass the CUDA stream to the output message
+  stream_handler_result = cuda_stream_handler_.to_message(out_mask_message);
   if (stream_handler_result != GXF_SUCCESS) {
     throw std::runtime_error("Failed to add the CUDA stream to the outgoing messages");
   }
 
-  auto result_host = gxf::Entity(std::move(out_message_host.value()));
-  auto result_device = gxf::Entity(std::move(out_message_device.value()));
+  auto result_host = gxf::Entity(std::move(out_coords_message.value()));
+  auto result_device = gxf::Entity(std::move(out_mask_message.value()));
   op_output.emit(result_host, "out_coords");
   op_output.emit(result_device, "out_mask");
```

Comment:
Originally these tensors were both on a single "out" port, but I [had to change to separate ports for the distributed version of the app](https://github.com/nvidia-holoscan/holohub/pull/262) because UCX cannot mix host and device tensors in the same entity transmitted by `UcxTransmitter` (the same UCX transport layer, e.g. cuda_copy, etc. has to be used for all tensors in a given entity)
```suggestion
  op_output.emit(result_host, "out_coords");
  op_output.emit(result_device, "out_mask");
```

Now that it looks like `out_coords` and `out_mask` are both in device memory, then it would be good to go back to having only a single "out" port again, which may be a bit more efficient, particularly in the distributed case as it reduces the number of separate messages that have to be sent between the fragments.

Reply:
Yes, makes sense. I'll change that back to a single output port.